serverlessworkflow/modules/ROOT/pages/security/token-exchange-for-openapi-services.adoc

```diff
@@ -0,0 +1,304 @@
+:configuring-openapi-services-endpoints: xref:service-orchestration/configuring-openapi-services-endpoints.adoc
+:authentication-support-for-openapi-services: xref:security/authention-support-for-openapi-services.adoc
+
+This guide shows how to configure OAuth 2.0 Token Exchange when invoking OpenAPI-secured services from workflows.
+
+For general OpenAPI orchestration and authentication, see:
+
+* {orchestration-of-openapi-based-services}[Orchestrating the OpenAPI services]
+* {configuring-openapi-services-endpoints}[Configuring the OpenAPI services endpoints]
+* {authentication-support-for-openapi-services}[Authentication for OpenAPI services]
+
+
+== When to use Token Exchange
+
+Use Token Exchange if:
+
+* You must avoid forwarding the original end-user token to third-party services.
+* You have long running workflow which may cause token invalidity (expiration).
+
+If you only need to forward the original token, see token propagation in {authentication-support-for-openapi-services}#ref-authorization-token-propagation[Authorization token propagation].
+
+== How Token Exchange works
+
+At a high level, Token Exchange lets the workflow swap the incoming bearer token for a new token tailored to the downstream service. The Quarkus OIDC Client Filter performs the exchange transparently before the OpenAPI client call.
+
+.Flow overview
+[%autowidth,cols="1,5"]
+|===
+|Step |Description
+
+|1 |Client calls the workflow REST endpoint and sends `Authorization: Bearer <user-access-token>`.
+|2 |The OpenAPI client (generated for a secured operation) is invoked by the workflow.
+|3 |The credential provider checks the cache for a valid exchanged token for this process instance and auth name; if found and not near expiry, it is reused.
+|4 |If no valid token is found, the OIDC Client Filter detects the `oauth2` security scheme and requests a Token Exchange at the IdP, using the incoming token as the subject token and the configured audience.
+|5 |The Identity Provider returns the exchanged access token (for example, audience `downstream-api`), which is cached with its expiry.
+|6 |The OpenAPI client calls the downstream service with `Authorization: Bearer <exchanged-token>`.
+|7 |Downstream service responds; workflow proceeds and returns the result to the client.
+|8 |Proactive refresh: a background monitor refreshes cached tokens nearing expiration based on `sonataflow.security.auth.<auth_name>.token-exchange.proactive-refresh-seconds` and `sonataflow.security.auth.token-exchange.monitor-rate-seconds`.
+|===
+
+.Sequence diagram
+link:images/security/token-exchange-sequence.svg[image:security/token-exchange-sequence.svg[width=100%],role="diagram"]
+
+== Requirements
+
+* Quarkus OIDC Client Filter extension in the workflow service:
+
+[source,xml]
+----
+<dependency>
+  <groupId>io.quarkus</groupId>
+  <artifactId>quarkus-oidc-client-filter</artifactId>
+</dependency>
```

Comment:
Please add a note about adding this dependency in the cloud space: https://sonataflow.org/serverlessworkflow/main/cloud/operator/build-and-deploy-workflows.html#passing-build-arguments-to-internal-workflow-builder

Reply:
done, I added the link everywhere there is a mention of add-on/extensions

Reply:
it did not click in my brain that it was a reference from the same doc :expressionless: